pennylane/templates/subroutines/trotter.py

```diff
@@ -0,0 +1,292 @@
+"""
+Contains templates for Suzuki-Trotter approximation based subroutines.
+"""
+import pennylane as qml
+from pennylane.operation import Operation
+from pennylane.ops import Sum
+
+
+def _scalar(order):
+    """Compute the scalar used in the recursive expression.
+
+    Args:
+        order (int): order of Trotter product (assume order is an even integer > 2).
+
+    Returns:
+        float: scalar to be used in the recursive expression.
+    """
+    root = 1 / (order - 1)
+    return (4 - 4**root) ** -1
+
+
+@qml.QueuingManager.stop_recording()
+def _recursive_expression(x, order, ops):
+    """Generate a list of operations using the
+    recursive expression which defines the Trotter product.
+
+    Args:
+        x (complex): the evolution 'time'
+        order (int): the order of the Trotter expansion
+        ops (Iterable(~.Operators)): a list of terms in the Hamiltonian
+
+    Returns:
+        list: the approximation as product of exponentials of the Hamiltonian terms
+    """
+    if order == 1:
+        return [qml.exp(op, x * 1j) for op in ops]
+
+    if order == 2:
+        return [qml.exp(op, x * 0.5j) for op in ops + ops[::-1]]
+
+    scalar_1 = _scalar(order)
+    scalar_2 = 1 - 4 * scalar_1
+
+    ops_lst_1 = _recursive_expression(scalar_1 * x, order - 2, ops)
+    ops_lst_2 = _recursive_expression(scalar_2 * x, order - 2, ops)
+
+    return (2 * ops_lst_1) + ops_lst_2 + (2 * ops_lst_1)
+
+
+class TrotterProduct(Operation):
+    r"""An operation representing the Suzuki-Trotter product approximation for the complex matrix
+    exponential of a given Hamiltonian.
+
+    The Suzuki-Trotter product formula provides a method to approximate the matrix exponential of
+    Hamiltonian expressed as a linear combination of terms which in general do not commute. Consider
+    the Hamiltonian :math:`H = \Sigma^{N}_{j=0} O_{j}`, the product formula is constructed using
+    symmetrized products of the terms in the Hamiltonian. The symmetrized products of order
+    :math:`m \in [1, 2, 4, ..., 2k]` with :math:`k \in \mathbb{N}` are given by:
+
+    .. math::
+
+        \begin{align}
+            S_{1}(t) &= \Pi_{j=0}^{N} \ e^{i t O_{j}} \\
+            S_{2}(t) &= \Pi_{j=0}^{N} \ e^{i \frac{t}{2} O_{j}} \cdot \Pi_{j=N}^{0} \ e^{i \frac{t}{2} O_{j}} \\
+            &\vdots \\
+            S_{m}(t) &= S_{m-2}(p_{m}t)^{2} \cdot S_{m-2}((1-4p_{m})t) \cdot S_{m-2}(p_{m}t)^{2},
+        \end{align}
+
+    where the coefficient is :math:`p_{m} = 1 / (4 - \sqrt[m - 1]{4})`. The :math:`m`th order,
+    :math:`n`-step Suzuki-Trotter approximation is then defined as:
+
+    .. math:: e^{iHt} \approx \left [S_{m}(t / n)  \right ]^{n}.
+
+    For more details see `J. Math. Phys. 32, 400 (1991) <https://pubs.aip.org/aip/jmp/article-abstract/32/2/400/229229>`_.
+
+    Args:
+        hamiltonian (Union[~.Hamiltonian, ~.Sum]): The Hamiltonian written as a linear combination
+            of operators with known matrix exponentials.
```

Comment:
Please remove the dot for consistency.

Reply:
I think if you remove the tilde before Ham and Sum it will link dynamically in the docs
https://xanaduai-pennylane--4661.com.readthedocs.build/en/4661/code/api/pennylane.TrotterProduct.html

Reply:
Removed "~", kept "."